File: JumpyOtter/Sources/MultiplayerClient.swift
import Foundation

/// Events the relay can deliver, surfaced to the game on an internal queue.
enum MultiplayerEvent {
    case connected(playerID: Int, peers: [Int])
    case peerJoined(id: Int)
    case peerState(id: Int, row: Int, x: Float, score: Int, alive: Bool)
    case peerGarbage(id: Int, amount: Int)
    case peerGameOver(id: Int, score: Int)
    case opponentLeft(id: Int)
    case disconnected
}

/// WebSocket client for the relay server (relay.py).
///
/// Exchanges small JSON messages: "state" (position + score, sent on every
/// change), "garbage" (hazard attack on rivals), "gameOver" (topped out) and
/// relay-generated "welcome" / "playerJoined" / "opponentLeft".
final class MultiplayerClient: NSObject {

    static let defaultURL = URL(string: "ws://127.0.0.1:8765")!

    private(set) var playerID: Int = -1
    private(set) var isConnected = false

    private let url: URL
    private var task: URLSessionWebSocketTask?
    private var session: URLSession!
    private var pingTimer: Timer?

    // Events are queued here and drained by the game on its render thread.
    private let eventLock = NSLock()
    private var pendingEvents: [MultiplayerEvent] = []

    init(url: URL = MultiplayerClient.defaultURL) {
        self.url = url
        super.init()
        session = URLSession(configuration: .default)
    }

    // MARK: - Connection

    func connect() {
        guard task == nil else { return }
        let task = session.webSocketTask(with: url)
        self.task = task
        task.resume()
        receiveLoop()
        DispatchQueue.main.async { [weak self] in
            self?.pingTimer = Timer.scheduledTimer(withTimeInterval: 15, repeats: true) { _ in
                self?.task?.sendPing { _ in }
            }
        }
    }

    func disconnect() {
        pingTimer?.invalidate()
        pingTimer = nil
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        isConnected = false
    }

    /// Drains queued events; call from the game loop.
    func drainEvents() -> [MultiplayerEvent] {
        eventLock.lock()
        let events = pendingEvents
        pendingEvents.removeAll()
        eventLock.unlock()
        return events
    }

    // MARK: - Outgoing messages

    func sendState(row: Int, x: Float, score: Int, alive: Bool) {
        sendJSON([
            "type": "state",
            "row": row,
            "x": Double(x),
            "score": score,
            "alive": alive,
        ])
    }

    func sendGarbage(amount: Int) {
        sendJSON(["type": "garbage", "amount": amount])
    }

    func sendGameOver(score: Int) {
        sendJSON(["type": "gameOver", "score": score])
    }

    private func sendJSON(_ dict: [String: Any]) {
        guard isConnected,
              let data = try? JSONSerialization.data(withJSONObject: dict),
              let text = String(data: data, encoding: .utf8) else { return }
        task?.send(.string(text)) { _ in }
    }

    // MARK: - Incoming messages

    private func receiveLoop() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let message):
                if case .string(let text) = message { self.handle(text) }
                self.receiveLoop()
            case .failure:
                if self.isConnected {
                    self.isConnected = false
                    self.enqueue(.disconnected)
                }
                self.task = nil
            }
        }
    }

    private func handle(_ text: String) {
        guard let data = text.data(using: .utf8),
              let msg = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = msg["type"] as? String else { return }
        let id = msg["id"] as? Int ?? -1
        switch type {
        case "welcome":
            playerID = id
            isConnected = true
            let peers = msg["players"] as? [Int] ?? []
            enqueue(.connected(playerID: id, peers: peers))
        case "playerJoined":
            enqueue(.peerJoined(id: id))
        case "state":
            enqueue(.peerState(
                id: id,
                row: msg["row"] as? Int ?? 0,
                x: Float(msg["x"] as? Double ?? 0),
                score: msg["score"] as? Int ?? 0,
                alive: msg["alive"] as? Bool ?? true))
        case "garbage":
            enqueue(.peerGarbage(id: id, amount: msg["amount"] as? Int ?? 1))
        case "gameOver":
            enqueue(.peerGameOver(id: id, score: msg["score"] as? Int ?? 0))
        case "opponentLeft":
            enqueue(.opponentLeft(id: id))
        default:
            break
        }
    }

    private func enqueue(_ event: MultiplayerEvent) {
        eventLock.lock()
        pendingEvents.append(event)
        eventLock.unlock()
    }
}

File: android/app/src/main/java/com/devin/jumpyotter/MultiplayerClient.kt
package com.devin.jumpyotter

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.TimeUnit

sealed class MultiplayerEvent {
    data class Connected(val playerID: Int, val peers: List<Int>) : MultiplayerEvent()
    data class PeerJoined(val id: Int) : MultiplayerEvent()
    data class PeerState(val id: Int, val row: Int, val x: Float, val score: Int, val alive: Boolean) : MultiplayerEvent()
    data class PeerGarbage(val id: Int, val amount: Int) : MultiplayerEvent()
    data class PeerGameOver(val id: Int, val score: Int) : MultiplayerEvent()
    data class OpponentLeft(val id: Int) : MultiplayerEvent()
    object Disconnected : MultiplayerEvent()
}

/**
 * WebSocket client for relay.py. Speaks the same JSON protocol as the iOS app.
 * Events are queued and drained on the game thread via [poll].
 */
class MultiplayerClient(private val url: String = DEFAULT_URL) {

    private val http = OkHttpClient.Builder()
        .connectTimeout(3, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()
    private var socket: WebSocket? = null
    private val events = ConcurrentLinkedQueue<MultiplayerEvent>()

    @Volatile var connected = false
        private set
    @Volatile var playerID = -1
        private set

    fun connect() {
        if (socket != null) return
        val request = Request.Builder().url(url).build()
        socket = http.newWebSocket(request, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) = handle(text)

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                connected = false
                socket = null
                events.add(MultiplayerEvent.Disconnected)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                connected = false
                socket = null
                events.add(MultiplayerEvent.Disconnected)
            }
        })
    }

    fun disconnect() {
        socket?.close(1000, null)
        socket = null
        connected = false
    }

    fun poll(): MultiplayerEvent? = events.poll()

    fun sendState(row: Int, x: Float, score: Int, alive: Boolean) {
        send(JSONObject().put("type", "state").put("row", row).put("x", x.toDouble()).put("score", score).put("alive", alive))
    }

    fun sendGarbage(amount: Int) {
        send(JSONObject().put("type", "garbage").put("amount", amount))
    }

    fun sendGameOver(score: Int) {
        send(JSONObject().put("type", "gameOver").put("score", score))
    }

    private fun send(obj: JSONObject) {
        if (!connected) return
        socket?.send(obj.toString())
    }

    private fun handle(text: String) {
        val obj = try { JSONObject(text) } catch (e: Exception) { return }
        when (obj.optString("type")) {
            "welcome" -> {
                playerID = obj.optInt("id", -1)
                connected = true
                events.add(MultiplayerEvent.Connected(playerID, intList(obj.optJSONArray("peers"))))
            }
            "peerJoined" -> events.add(MultiplayerEvent.PeerJoined(obj.optInt("id")))
            "state" -> events.add(
                MultiplayerEvent.PeerState(
                    obj.optInt("id"), obj.optInt("row"), obj.optDouble("x", 0.0).toFloat(),
                    obj.optInt("score"), obj.optBoolean("alive", true),
                )
            )
            "garbage" -> events.add(MultiplayerEvent.PeerGarbage(obj.optInt("id"), obj.optInt("amount", 1)))
            "gameOver" -> events.add(MultiplayerEvent.PeerGameOver(obj.optInt("id"), obj.optInt("score")))
            "opponentLeft" -> events.add(MultiplayerEvent.OpponentLeft(obj.optInt("id")))
        }
    }

    private fun intList(arr: JSONArray?): List<Int> {
        if (arr == null) return emptyList()
        return (0 until arr.length()).map { arr.optInt(it) }
    }

    companion object {
        /** 10.0.2.2 is the emulator's alias for the host loopback, where relay.py listens. */
        const val DEFAULT_URL = "ws://10.0.2.2:8765"
    }
}
